api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,324 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+
+1. Get the code with line numbers from the current editor.
+2. Send the code to the Language Model API to get the annotations.
+3. Parse the annotations and display them in the editor.
+
+Let's start by getting the code from the current editor. Add the following method directly above the `export function deactivate() { }` line at the bottom of the `extension.ts` file.
+
+```ts
+function getCodeWithLineNumbers(textEditor: vscode.TextEditor) {
+ let currentLine = textEditor.visibleRanges[0].start.line;
+ const endLine = textEditor.visibleRanges[0].end.line;
+ let code = '';
+ while (currentLine < endLine) {
+  code += `${currentLine + 1}: ${textEditor.document.lineAt(currentLine).text} \n`;
+  currentLine++;
+ }
+ return code;
+}
+```
+
+This code uses the `visibleRanges` property of the TextEditor to get the lines that are currently visible in the editor. It then loops through these lines and adds each one to a string, along with the line number. Finally, it returns the string that contains the code with line numbers.
+
+Now we can call this method from the `code-tutor.annotate` command. Modify the implementation of the command so that it looks like this:
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getCodeWithLineNumbers(textEditor);
+
+});
+```
+
+The next step is to call the GitHub Copilot language model. To do this, we first need to specify which chat model we want to use.
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+});
+```
+
+Now we need to send the code to the language model along with a prompt that tells the model to generate annotations for the code. Add the following code to the top of the file directly under the imports.
+
+```ts
+const ANNOTATION_PROMPT = `You are a code tutor who helps students learn how to write better code. Your job is to evaluate a block of code that the user gives you. The user is writing You will then annotate any lines that could be improved with a brief suggestion and the reason why you are making that suggestion. Only make suggestions when you feel the severity is enough that it will impact the readibility and maintainability of the code. Be friendly with your suggestions and remember that these are students so they need gentle guidance. Format each suggestion as a single JSON object. It is not necessary to wrap your response in triple backticks. Here is an example of what your response should look like:
```

Comment:
I would call this SYSTEM_PROMPT, I think this is the standard language that the ai community uses.